[Wallet] Historical currency conversions in the transaction feed

jeanregisser opened the pull request:
### Description

This PR enables historical currency conversions in the transaction feed.

Main highlights:
- Implemented the updated schema in `blockchain-api` which allows querying transactions only for a specific token (cUSD on the home screen, cGLD on the exchange screen), retrieving
amounts in local currency, and later on will make it easy to paginate the data as proposed in https://docs.google.com/document/d/1ryfz7Gd0bk9QQHDBCWyaBy2AeOui63hgp-HdFtyBTyc/edit
- Fixed the broken `gql-gen` tool and made GraphQL related code type safe again by using `graphql-codegen` for the client. (we should use it for `blockchain-api` too)
- Started moving displayed money amounts into the `CurrencyDisplay` component. This component takes care of doing the actual currency conversions when needed. It will evolve as we move more of the existing code to it. It already simplified the feed items components.
- Refactored code to use the same `TransactionsList` component for both the home and exchange screens and added tests for it.

We should probably run `graphql-codegen` as part of our React Native dev startup script to spot any type issues with the `blockchain-api` we're connecting to. I'll address that in a separate PR.

In the `ExchangeFeedItem` fragment, the `amount` field might feel redundant as it's either the `makerAmount` or the `takerAmount` with the addition of the sign (+/-), which we could derive client side. My main goal was to make the interface right first and optimize later when needed.

The `TransactionsList.text.tsx` test outputs console logs about invalid/missing fragments, this is caused by the Apollo `MockProvider` context missing details about the `interface` type we're using. I tried to pass the right `fragmentMatcher` like we do in the normal client (see https://www.apollographql.com/docs/react/data/fragments/#fragments-on-unions-and-interfaces)
but it then failed to return any data to the querying component.
In short, the test works as expected but outputs a lot of misleading noise.

Finally, this PR looks a bit big, but there are a lot of simple renames and snapshots changes, so hopefully shouldn't be too hard to review 🙏 

### Tested

- Confirmed that converted local amounts in the transaction feed now use historical rates.
- Added new tests.

### Other changes

- Currency conversion API now outputs a `string` instead of a `number` to prevent rounding issues.
- The amounts in the transactions API are now `string` as well for the same reason stated above.
- Memoized gold exchange rates queries (per requests).
- Fixed a redux state update that was done in a non immutable way.
- I had a problem with the test suite when running with `--runInBand` like in the CI and it was caused by the apollo query component not being unmounted after the test.
I upgraded `react-native-testing-library` to be able to cleanup mounted components after testing. See https://github.com/callstack/react-native-testing-library/blob/master/docs/API.md#cleanup
I decided to make that behavior global for all tests as I feel it's the right thing to do for us.
See https://github.com/celo-org/celo-monorepo/pull/2446/commits/b99484a90b6465ffcbbdddf5c1fb8585902322d8
I just had to change one test which relied on a component being mounted between test.
See https://github.com/celo-org/celo-monorepo/pull/2446/commits/6a9031067ad2f89c428c6f917dce57b7f36b4e6d
Let me know if you have any concern with this.

### Related issues

- Fixes #834
- Fixes #1386
- #1261
- Starts addressing #1484

### Backwards compatibility

Yes.

jeanregisser committed:
[WiP] New schema to support historical rates in the transaction feed

jeanregisser committed:
[WiP] Fix types and restore functionalities (except standby transaction)

jeanregisser committed:
Revert "[Wallet] Fix slow activity feed tx confirmations (#2290)"

This reverts commit 8ebc6633f34f3916a6e25faa75378eccba9cdfe2.

Conflicts too much with the transaction feed refactor.

jeanregisser committed:
Restore better log message

jeanregisser committed:
[WiP] Map standby transactions to the new types for rendering with local amounts

jeanregisser committed:
[WiP] Use MoneyAmount type in the exchange confirmation card

jeanregisser committed:
[WiP] Use MoneyAmount type in the transfer confirmation card

jeanregisser committed:
[WiP] Make CurrencyDisplay handle more currency display use cases

jeanregisser committed:
[WiP] Simplify ExchangeFeedItem by using CurrencyDisplay

jeanregisser committed:
Fix non immutable redux state update

jeanregisser committed:
Memoize gold exchange rates requests (per request)

jeanregisser committed:
Add test for TransactionsList

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/transaction-feed-refactor

jeanregisser committed:
Cleanup code

jeanregisser committed:
Fix tests

i1skn commented on packages/blockchain-api/src/blockscout.ts:
```diff
@@ -214,6 +218,107 @@ export class BlockscoutAPI extends RESTDataSource {
+        // Find the event related to the queried token
+        const tokenEvent = [inEvent, outEvent].find((event) => event.tokenSymbol === args.token)
+        if (tokenEvent) {
+          const timestamp = new BigNumber(inEvent.timeStamp).toNumber()
```
would `parseInt` do the job here?

jmrossy replied:
We generally opt for BigNum over parseInt

i1skn replied:
Did not know that. Is the reason because `BigNumber` is more universal and can handle different input formats compare to `parseInt`?

jmrossy replied:
exactly

i1skn commented on packages/blockchain-api/src/currencyConversion/GoldExchangeRateAPI.ts:
```diff
@@ -56,7 +57,19 @@ export default class GoldExchangeRateAPI<TContext = any> extends DataSource {
     const pair = `${sourceCurrencyCode || CUSD}/${currencyCode}`
+    const cacheKey = `${pair}-${date.getTime()}`
 
+    let promise = this.memoizedResults.get(cacheKey)
```
What do you think about:
```
    let promise = this.memoizedResults.get(cacheKey)
    if (!promise) {
       promise = this.performRequest(pair, date)
       this.memoizedResults.set(cacheKey, promise)
    }
    return promise
```
?

jeanregisser replied:
Yep better thanks!

i1skn commented on packages/mobile/src/send/TransferConfirmationCard.tsx:
```diff
@@ -67,56 +60,47 @@ const renderTopSection = (props: Props) => {
+          type={DisplayType.Big}
+          amount={amount}
+          // tslint:disable-next-line: jsx-no-lambda
+          formatAmount={(value) => getNetworkFeeDisplayValue(value, true)}
```
Is there particular reason to use lambda function here?

jeanregisser replied:
No, let me change this 👍

i1skn commented on packages/mobile/src/transactions/TransactionFeed.tsx:
```diff
@@ -38,137 +28,95 @@ interface StateProps {
+    const { addressToE164Number, invitees, recipientCache } = this.props
+
+    if (tx.hasOwnProperty('comment')) {
+      // @ts-ignore
```
Is it possible to use`if (tx.__typename === 'TransactionTransfer') {` to avoid `// @ts-ignore` here?

jeanregisser replied:
Yes, will change it 👍

i1skn commented on packages/mobile/src/transactions/TransactionsList.tsx:
```diff
@@ -0,0 +1,291 @@
+
+const mapStateToProps = (state: RootState): StateProps => ({
+  address: currentAccountSelector(state),
+  standbyTransactions: state.transactions.standbyTransactions,
```
what is the convention about selectors, do we create them only when some logic is involved to encapsulate it?

jeanregisser replied:
I usually introduce them when it's needed in more than 2 places. Or If I need reselect.

i1skn commented on packages/mobile/src/transactions/TransactionsList.tsx:
```diff
@@ -0,0 +1,291 @@
+    return (
+      <UserTransactionsComponent
+        query={TRANSACTIONS_QUERY}
+        pollInterval={10000}
```
Can we move it to `const`?

jeanregisser replied:
👍

i1skn commented:
Great work!!! See a couple of non-blocker comments I've left :)

jeanregisser committed:
Address review comments

jeanregisser committed:
Upgrade react-native-testing-library to be able to cleanup mounted trees

This fixes the TransactionsList test preventing the test suite to stop cleanly as the polling query was still mounted and doing its work.

jeanregisser committed:
Fix InviteReview test now that components are unmounted after each test

jeanregisser committed:
Remove noisy debug logs now that we do conversions for each transaction

jeanregisser committed:
Use milliseconds timestamp

i1skn reviewed (approved):
Thanks for the tremendous improvement 🚢 !

jeanregisser committed:
Add test for blockscout getTransactions

jeanregisser committed:
Hide currency symbol in the exchange item details

jeanregisser committed:
Add depreciation notice

jeanregisser committed:
Add description for custom scalars

jeanregisser committed:
Better description for Address courtesy of https://github.com/poanetwork/blockscout/blob/130a4084eaf2c0db6ede18266b6f02e6a9230163/apps/block_scout_web/lib/block_scout_web/schema/scalars.ex#L11-L16

jeanregisser committed:
Remove unneeded prop

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/transaction-feed-refactor

jmrossy commented on packages/blockchain-api/src/blockscout.ts:
```diff
@@ -214,6 +218,107 @@ export class BlockscoutAPI extends RESTDataSource {
     return rewards.sort((a, b) => b.timestamp - a.timestamp)
   }
+
+  async getTransactions(args: TransactionArgs) {
```
Can you leave a comment or rename this to clarify how it differs from the existing `getTokenTransactions` and  `getFeedEvents` methods?

jeanregisser replied:
Yes! `getFeedEvents` was kept for backward compatibility, it is intended to be removed at some point.

jmrossy replied:
Ok. Up to you then if you want to remove it now or later. Please leave TODOs explaining this

jmrossy commented on packages/blockchain-api/src/blockscout.ts:
```diff
@@ -214,6 +218,107 @@ export class BlockscoutAPI extends RESTDataSource {
+      txHashToEventTransactions.set(tx.hash, currentTX)
+    }
+
+    await this.ensureTokenAddresses()
```
Not a new change in this PR but we should consider reworking this bit or at least leave a comment. This design of forcing users of token addresses to first call this method to ensure they're set is problematic. Easy to mess up. Would be better if the token addresses had getters that fetched them when null. cc @annakaz

jeanregisser replied:
Also I suspect this call is slowing down queries given a new data source is instantiated for each request.

jmrossy commented on packages/blockchain-api/src/blockscout.ts:
```diff
@@ -214,6 +218,107 @@ export class BlockscoutAPI extends RESTDataSource {
+
+    await this.ensureTokenAddresses()
+    // Generate final events
+    txHashToEventTransactions.forEach((transactions: BlockscoutTransaction[], txhash: string) => {
```
Can we DRY this up with the `getFeedEvents` method above? Or is one being deprecated?

jeanregisser replied:
yes `getFeedEvents` is being deprecated.

jmrossy commented on packages/blockchain-api/src/schema.ts:
```diff
@@ -106,26 +217,57 @@ export const typeDefs = gql`
     rewards: async (_source: any, args: EventArgs, { dataSources }: Context) => {
       return dataSources.blockscoutAPI.getFeedRewards(args)
     },
+    transactions: async (_source: any, args: TransactionArgs, context: Context) => {
```
My problem with `transaction` is that it's too vague. It could be anything from an ERC20 token transfer to a verification request to a smart contract deployment. Is there a more descriptive name we could use?

jeanregisser replied:
Makes sense, `tokenTransactions` feels like the right name then, what do you think?

jmrossy replied:
Sounds good

jeanregisser replied:
done 👍

jmrossy commented on packages/mobile/src/apollo/types.ts:
```diff
@@ -1,54 +1,328 @@
-  comment: string
-  status?: TransactionStatus
-  // TODO: fee needs to be added here
+export type Event = Exchange | Transfer
```
I guess you kept the old Event interface for backwards compatibility with older app versions? But the next time we ship the app, we'll be resetting the network too and forcing users to update. We could actually remove all that.

jeanregisser replied:
Indeed! Let's remove it then.

jeanregisser replied:
I'll remove the deprecated code after this PR is merged to prevent breaking things unnecessarily given `blockchain-api` has to be redeployed.

jmrossy commented on packages/mobile/src/components/CurrencyDisplay.tsx:
```diff
@@ -2,61 +2,128 @@ import colors from '@celo/react-components/styles/colors'
+  size: number // only used for DisplayType.Big
+  useColors: boolean
+  hideSymbol: boolean
+  formatAmount: (amount: BigNumber.Value) => string
```
Can the formatAmount be determined from the DisplayType? Or maybe made optional? Just wondering if we can keep as much currency formatting related logic encapsulated in this file as possible.

jeanregisser replied:
Though it's not visible here, it's optional (see `defaultProps`) and used to format some network fee amounts. 
DisplayType is currently used for determining how we display the symbol (superscript vs normal, maybe I should rename that too).
I thought about providing another enum for the format, but decided to leave it as is with a default implementation until I refactor more of the code related to local currencies.
I agree it should encapsulate all formatting logic too.

jmrossy commented on packages/mobile/src/components/CurrencyDisplay.tsx:
```diff
@@ -2,61 +2,128 @@ import colors from '@celo/react-components/styles/colors'
-  currencySymbol?: LocalCurrencySymbol | null
+  type: DisplayType
+  amount: MoneyAmount
+  size: number // only used for DisplayType.Big
```
what is size?

jeanregisser replied:
It's the fontSize, somewhat redundant with the style property. I want to address that in a follow up PR.

jmrossy commented on packages/mobile/src/components/CurrencyDisplay.tsx:
```diff
@@ -2,61 +2,128 @@ import colors from '@celo/react-components/styles/colors'
-      transform: [{ translateY: Math.round(size * 0.1) }],
-    }
-  }
+// TODO(Rossy) This is mostly duped by MoneyAmount, converge the two
```
Does this PR solve this TODO?

jeanregisser replied:
Not entirely, I'll move existing `MoneyAmount` usage in a follow up PR related to a  more holistic solution for displaying currencies.

jmrossy commented on packages/mobile/src/components/CurrencyDisplay.tsx:
```diff
@@ -2,61 +2,128 @@ import colors from '@celo/react-components/styles/colors'
+  style,
+}: Props) {
+  const localCurrencyCode = useLocalCurrencyCode()
+  // tslint:disable-next-line: react-hooks-nesting
```
why is this triggering?

jeanregisser replied:
It's because of the conditional at the end `|| xxx`.

jmrossy commented on packages/mobile/src/components/CurrencyDisplay.tsx:
```diff
@@ -2,61 +2,128 @@ import colors from '@celo/react-components/styles/colors'
+  )
+}
+
+CurrencyDisplay.defaultProps = {
```
If we're setting these default props, should these values be made optional with `?` in the interface above?

jeanregisser replied:
TypeScript does this automatically when consuming this component from JSX.

jmrossy commented on packages/mobile/src/components/CurrencyDisplay.tsx:
```diff
@@ -2,61 +2,128 @@ import colors from '@celo/react-components/styles/colors'
 
 const styles = StyleSheet.create({
-  container: {
+  container: {},
```
nit : remove

jmrossy commented on packages/mobile/src/exchange/ExchangeConfirmationCard.tsx:
```diff
@@ -42,7 +33,7 @@ const getExchangeRate = (props: Props) => {
 
   // For feed drilldown, the exchange rate has not been provided
-  return makerAmount.dividedBy(takerAmount)
+  return new BigNumber(makerAmount.amount).dividedBy(takerAmount.amount)
```
Per the thread you started on Slack, would be cool if we could encapsulate more logic in MoneyAmount and avoid this pattern :)

jeanregisser replied:
Yes we can definitely explore that.
I like the simplicity of having plain objects and functions operating on them. But money amount is an essential block in our app and we might benefit encapsulating it in its own class 👍

jmrossy commented on packages/mobile/src/firebase/notifications.ts:
```diff
@@ -59,17 +59,20 @@ function* handlePaymentReceived(
-        value: divideByWei(transferNotification.value),
-        currency: resolveCurrency(transferNotification.currency),
+        amount: {
+          amount: divideByWei(transferNotification.value).toString(),
```
Another common source of confusion is whether or not a value is in wei or not. You don't need to fix in this PR but would be cool if we had a wei to type that and avoid potential mistakes.

jeanregisser replied:
Yep, I will definitely continue thinking about this as we improve currencies handling.

jmrossy commented on packages/mobile/src/redux/migrations.test.ts:
```diff
@@ -68,4 +68,16 @@ describe('Redux persist migrations', () => {
     expect(migratedSchema.identity.hasSeenVerificationNux).toBeTruthy()
   })
+
+  it('work for v4 to v5', () => {
```
I don't believe v4 has shipped to users yet so no need for a v5. But it's fine either way.
In fact, we should blow these away and start fresh before the next deployment because we can only do that when the update coincides with a network reset.

jeanregisser replied:
Ok yes I'll leave them and we can reset all of these in another PR.

jmrossy commented on packages/mobile/src/transactions/ExchangeFeedItem.tsx:
```diff
@@ -4,161 +4,44 @@ import colors from '@celo/react-components/styles/colors'
-
+  const icon =
+    makerAmount.currencyCode === CURRENCIES[CURRENCY_ENUM.GOLD].code
+      ? require('src/transactions/ExchangeGoldGreen.png')
```
please move these `require` calls to src/images/Images.ts and then use the exported values here

jmrossy commented on packages/mobile/src/transactions/TransactionsList.test.tsx:
```diff
@@ -0,0 +1,195 @@
+import * as React from 'react'
```
Nice testing here 🥇

jmrossy commented on packages/mobile/src/transactions/TransactionsList.tsx:
```diff
@@ -0,0 +1,310 @@
+type Props = OwnProps & StateProps & DispatchProps
+
+// See https://github.com/microsoft/TypeScript/issues/16069#issuecomment-565658443
+function isPresent<T>(t: T | undefined | null | void): t is T {
```
Consider moving this to a shared place. Like utils/typescript.ts

jeanregisser replied:
Yep!

jmrossy reviewed (commented):
Overall really nice work! Some comments below but nothing major

jeanregisser committed:
Address some comments

jeanregisser committed:
Merge remote-tracking branch 'origin/master' into jeanregisser/transaction-feed-refactor

jeanregisser committed:
Rename MoneyAmount field `amount`-> `value`

jeanregisser committed:
Rename transactions -> tokenTransactions

Commit:
Merge refs/heads/master into jeanregisser/transaction-feed-refactor

Commit:
Merge refs/heads/master into jeanregisser/transaction-feed-refactor

jeanregisser committed:
Merge branch 'master' into jeanregisser/transaction-feed-refactor
